nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -0,0 +1,524 @@
+            loss_tensor = torch.concat(loss_tensors_list)
+            loss_mean = loss_tensor.mean()
+            acc_tensors_list = [loss_reduced["acc"] for loss_reduced in losses_reduced_per_micro_batch]
+
+            if len(acc_tensors_list) == 1:
+                acc_tensor = acc_tensors_list[0]
+            elif len(acc_tensors_list) > 1:
+                acc_tensor = torch.concat(acc_tensors_list)
+            acc_mean = acc_tensor.mean()
+        else:
+
+            loss_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            acc_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+
+            rewards_chosen_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_rejected_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_std = torch.tensor(0.0, device=torch.cuda.current_device())
+
+        # we can only log on one rank if it is rank zero so we broadcast from last rank
+        torch.distributed.broadcast(loss_mean, get_last_rank())
+        torch.distributed.broadcast(acc_mean, get_last_rank())
+
+        torch.distributed.broadcast(rewards_chosen_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_rejected_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_std, get_last_rank())
+
+        metrics = {
+            "loss": loss_mean,
+            "acc": acc_mean,
+            "rewards_actual_mean": rewards_chosen_mean,
+            "rewards_generated_mean": rewards_rejected_mean,
+            "rewards_all_mean": rewards_all_mean,
+            "rewards_all_std": rewards_all_std,
+        }
+
+        # move to CPU
+        metrics = {k: v.item() for k, v in metrics.items()}
+
+        return loss_mean.item(), metrics
+
+    def get_loss_and_metrics_vanilla_sft(self, batch, forward_only):
+        """Take a data_iter which is an interator over the microbatches
+            and return loss as well as metrics
+        """
+        seq_length = batch["tokens"].shape[1]
+        batch = {k: v for k, v in batch.items() if isinstance(v, torch.Tensor)}
+
+        data_iter = get_iterator_k_split(batch, get_num_microbatches())
+
+        set_sync_funcs(self, forward_only)
+
+        orig_loss_func = self.loss_func
+        self.loss_func = super().loss_func
+
+        fwd_bwd_function = get_forward_backward_func()
+        fwd_loss_fn = super().get_forward_output_and_loss_func(forward_only)
+
+        losses_reduced = fwd_bwd_function(
+            forward_step_func=fwd_loss_fn,
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=get_num_microbatches(),
+            forward_only=forward_only,
+            micro_batch_size=self.cfg.micro_batch_size,
```

Comment:
Shouldn't the micro-batch size be taken from `validation_ds.micro_batch_size`? (unless we get rid of that one)

Reply:
Correct, yes, good catch, I have changed everything now to be in line with how SFT does it